packages/core/src/plugins/html-deserializer/utils/deserializeHtml.spec.tsx

```diff
@@ -232,3 +232,36 @@ describe('when deserializing all plugins', () => {
     ).toEqual(output.children);
   });
 });
+
+describe('when stripWhitespace is true', () => {
+  // https://github.com/udecode/plate/issues/2713#issuecomment-1780118687
+  const html = `<p>\n  Hello  world\n  </p>\n\n  <p>\n  one     two   \n  three\n</p>\n\n<pre>\nhello     one two\nthree\nfour\n</pre>\n\n<div style="white-space: pre">\nhello     one two\nthree\nfour\n</div>\n\n<div style="white-space: pre-line">\nhello     one two\nthree\nfour\n</div>`;
+  const element = getHtmlDocument(html).body;
+
+  const expectedOutput = [
+    {
+      text: 'Hello world',
+    },
+    {
+      text: 'one two three',
+    },
+    {
+      text: 'hello one two three four',
+    },
+    {
+      text: '\nhello     one two\nthree\nfour\n',
+    },
+    {
+      text: '\nhello one two\nthree\nfour\n',
```

Comment:
Curiously, Firefox parses these with leading spaces, but no trailing spaces.

```suggestion
      text: '\nhello     one two\nthree\nfour',
    },
    {
      text: '\nhello one two\nthree\nfour',
```

Reply:
The behavior I observed in Firefox does not match you expected results. In my case, all the `\n` characters are preserved, and multiple `\n` characters at the end of lines are not merged into one (Windows 10). If you think it's necessary to remove trailing \n characters, you can simply add `replaceAll(/\n+$/g, '')`. However, I'm unsure if this approach will produce the desired results in other browsers.

Reply:
Huh, that's interesting. I'm on macOS Firefox. Let me check in a Windows VM to see if there really is an OS difference...

Reply:
Here's what I'm seeing when opening the plain HTML file in both Firefox and Chrome on macOS:

https://github.com/udecode/plate/assets/4272090/be75e846-7a26-4e7e-807a-0d0dc467c058

Reply:
I'm seeing the same thing on Windows 11 Firefox: leading newlines and no trailing newlines on both the divs, nothing on the pre.

https://github.com/udecode/plate/assets/4272090/db9bc5f7-43bb-4506-9d9a-70b4cdcfd67a

Reply:
You can try adding multiple \n characters at the end of a line.

Reply:
Ah, I see what's happening. The rule seems to be:

- For `<pre>` tags
  - Strip exactly one `\n` from the start of the text node
  - Strip exactly one `\n` from the end of the text node
- For `white-space: pre or pre-line`:
  - Do not strip any `\n` from the start of the text node
  - Strip exactly one `\n` from the end of the text node

Reply:
If this fits with your observations, can you implement this in your solution and add test cases to verify it?

Reply:
Yes, I have noticed the same behavior. However, I am not familiar with the specific specifications, so I did not make any modifications.

Reply:
To be honest, I couldn't make too much sense of the specification either. 😅

The part about `<pre>` tags treating leading newlines differently is documented [here](https://html.spec.whatwg.org/#the-pre-element), at least.

The overall goal of this endeavour is to make sure that when a user pastes content from a browser into Plate, the resulting content is spaced the same way as it was in the browser, so if browsers are doing things one way, that's the way we should be doing them, even if the spec behind it is difficult to track down or make sense of.

Reply:
> Ah, I see what's happening. The rule seems to be:
> 
> * For `<pre>` tags
>   
>   * Strip exactly one `\n` from the start of the text node
>   * Strip exactly one `\n` from the end of the text node
> * For `white-space: pre or pre-line`:
>   
>   * Do not strip any `\n` from the start of the text node
>   * Strip exactly one `\n` from the end of the text node

I added the expected behavior for the `<pre>` tag, but I found that it is being overridden by the `code block`.

Reply:
Thanks! I'll take a look at the code block behaviour separately; consider it out of scope for this PR. 🙂